lib/record_store/provider/ns1/client.rb

```diff
@@ -1,46 +1,58 @@
+require 'net/http'
 require 'ns1'
 
 module RecordStore
   class Provider::NS1 < Provider
-    class Error < StandardError; end
-
     class Client < ::NS1::Client
       def initialize(api_key:)
         super(api_key)
       end
 
       def zones
-        super
+        zones = super
+        raise_error(zones) if zones.is_a?(NS1::Response::Error)
+        zones
       end
 
       def zone(name)
-        super(name)
+        zone = super(name)
+        raise_error(zone) if zone.is_a?(NS1::Response::Error)
+        zone
       end
 
       def record(zone:, fqdn:, type:, must_exist: false)
         result = super(zone, fqdn, type)
-        raise(Error, result.to_s) if must_exist && result.is_a?(NS1::Response::Error)
+        raise_error(result) if must_exist && result.is_a?(NS1::Response::Error)
         return nil if result.is_a?(NS1::Response::Error)
         result
       end
 
       def create_record(zone:, fqdn:, type:, params:)
         result = super(zone, fqdn, type, params)
-        raise(Error, result.to_s) if result.is_a?(NS1::Response::Error)
+        raise_error(result) if result.is_a?(NS1::Response::Error)
         nil
       end
 
       def modify_record(zone:, fqdn:, type:, params:)
         result = super(zone, fqdn, type, params)
-        raise(Error, result.to_s) if result.is_a?(NS1::Response::Error)
+        raise_error(result) if result.is_a?(NS1::Response::Error)
         nil
       end
 
       def delete_record(zone:, fqdn:, type:)
         result = super(zone, fqdn, type)
-        raise(Error, result.to_s) if result.is_a?(NS1::Response::Error)
+        raise_error(result) if result.is_a?(NS1::Response::Error)
         nil
       end
+
+      private
+
+      def raise_error(result)
+        if Net::HTTPResponse::CODE_TO_OBJ[result.status.to_s] == Net::HTTPServiceUnavailable
+          raise RecordStore::Provider::ProviderUnavailableError, result.to_s
+        end
```

Comment:
This is nice! To iterate on this, I'm wondering if we could maybe have an `UnparsableError` type also that would be raised for every error type that has content not parsable through JSON ? It would be a matter of trying to parse the error content and raise if not possible. That way we do not have to add exceptions for every type of error, and an unparsable error should always be a service issue and not an API error, so we know it's worth retrying!